### PR TITLE
[TT-3904] fix chunked body logging

### PR DESCRIPTION
**Pull request by urbanishimwe**
when request/response body is sent in chunks, log browser doesn't display the body.

## Description
check the description of the linked issue

## Related Issue
https://tyktech.atlassian.net/browse/TT-3904

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

### CODE REVIEW

File: gateway/handler_success.go [hunk 1]
```diff
@@ -352,3 +347,10 @@ func (s *SuccessHandler) ServeHTTPWithCache(w http.ResponseWriter, r *http.Reque
 
 	return inRes
 }
+
+func requestBodyGraphQL(r *http.Request) {
+	if gqlCtx := ctxGetGraphQLRequest(r); gqlCtx != nil {
+		contents, _ := json.Marshal(gqlCtx)
+		r.Body = nopCloser{bytes.NewReader(contents)}
+	}
+}
```
Comment (vaske) [inline gateway/handler_success.go:356]: missing empty line on file end

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3861/merge
Commit:  
Triggered by: pull_request (@urbanishimwe)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/1744487252)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3861/merge
Commit: e29513a5cbb660b3924bcb9f28a82457f4f2f715 
Triggered by: pull_request (@vaske)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/1744552160)

File: gateway/handler_success.go [hunk 2]
```diff
@@ -352,3 +347,10 @@ func (s *SuccessHandler) ServeHTTPWithCache(w http.ResponseWriter, r *http.Reque
 
 	return inRes
 }
+
+func requestBodyGraphQL(r *http.Request) {
+	if gqlCtx := ctxGetGraphQLRequest(r); gqlCtx != nil {
```
Comment (furkansenharputlu) [inline gateway/handler_success.go:352]: Why is that related to GraphQL requests?

File: gateway/handler_success.go [hunk 3]
```diff
@@ -170,17 +171,11 @@ func (s *SuccessHandler) RecordHit(r *http.Request, timing Latency, code int, re
 			// mw_redis_cache instead? is there a reason not
 			// to include that in the analytics?
 			if responseCopy != nil {
-				contents, err := ioutil.ReadAll(responseCopy.Body)
```
Comment (furkansenharputlu) [inline gateway/handler_success.go:173]: This code change causes `TestRedisCacheMiddleware_WithCompressedResponse` to fail.

File: gateway/handler_success.go (hunk 2 above)
Comment (urbanishimwe) [reply gateway/handler_success.go:352]: GraphQL middleware reads the request body and unmarshal it inside *gql.Request which is stored in the request context. @furkansenharputlu

Comment (urbanishimwe) [reply gateway/handler_success.go:352]: @furkansenharputlu there's something about that function that make it "mis-behave" for chunked body.
so the way to fix it is to comment these lines:
https://github.com/TykTechnologies/tyk/blob/master/gateway/reverse_proxy.go#L1604-L1606
https://github.com/TykTechnologies/tyk/blob/master/gateway/reverse_proxy.go#L1617-L1619

but it seems they mean something

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3861/merge
Commit: 66d6cb0bea9a082f1f35e5ee5d853cd45079f39a 
Triggered by: pull_request (@urbanishimwe)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/1749873589)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3861/merge
Commit: 1da349892abe3b8aae55c31f69ad3d75e75f980a 
Triggered by: pull_request (@urbanishimwe)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/1756184327)

**Review by jeffy-mathew — APPROVED** [review]

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3861/merge
Commit: a14107cb6e5a5a069835e0eb0d5d9c3533e6ebb0 
Triggered by: pull_request (@urbanishimwe)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/1761510914)

**Review by furkansenharputlu — APPROVED** [review]

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3861/merge
Commit: 821018318608cc58e4f890dac48a84b017dcaf75 
Triggered by: pull_request (@urbanishimwe)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/1761694241)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3861/merge
Commit: bbfe7a0ae1ae88fca80a6e3b00ee0e26d2693a18 
Triggered by: pull_request (@urbanishimwe)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/1761702783)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/heads/master
Commit: bedf03a3d10f3c704738bdf9449c430bdbb719ca [TT-3904] fix chunked body logging (#3861)

* Fix chunked body recording in log browser
[changelog]
Fixed: reading response body twice, missing body for GraphQL request.

* fix tests

* fix gRPC tests

* add some test

* fixed lint error
Triggered by: push (@urbanishimwe)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/1762236403)
